crates/rpc/src/traits.rs

```diff
@@ -0,0 +1,65 @@
+use crate::api::SupervisorApiClient;
+use alloc::{boxed::Box, vec::Vec};
+use alloy_primitives::Log;
+use alloy_sol_types::SolEvent;
+use async_trait::async_trait;
+use core::time::Duration;
+use maili_interop::{ExecutingMessage, SafetyLevel, CROSS_L2_INBOX_ADDRESS};
+use tokio::time::error::Elapsed;
+
+/// Failures occurring during validation of [ExecutingMessage]s.
+#[derive(thiserror::Error, Debug)]
+pub enum ExecutingMessageValidatorError {
+    /// Failure during Supervisor's validation of [ExecutingMessage]s.
+    #[error("Supervisor determined messages are invalid: {0}")]
+    SupervisorValidationError(#[from] jsonrpsee_core::ClientError),
+
+    /// Message validation against the Supervisor took longer than allowed.
+    #[error("Message validation timed out: {0}")]
+    ValidationTimeout(#[from] Elapsed),
+}
+
+/// Interacts with a Supervisor to validate [ExecutingMessage]s.
+#[async_trait]
+pub trait ExecutingMessageValidator {
+    /// RPC client to Supervisor instance used for [ExecutingMessage] validation.
+    type SupervisorClient: SupervisorApiClient + Sync;
+
+    /// Default duration that message validation is not allowed to exceed.
+    const DEFAULT_TIMEOUT: Duration;
+
+    /// Extracts [ExecutingMessage]s from the [Log] if there are any.
```

Comment:
```suggestion
    /// Extracts [`ExecutingMessage`]s from the [`Log`]s if there are any.
```
nitpick, let's keep code with code syntax